Sources/NIOFileSystem/Internal/SystemFileHandle.swift

```diff
@@ -1116,17 +1116,28 @@ extension SystemFileHandle {
         if delayMaterialization {
             // When opening in this mode we can more "atomically" create the file, that is, by not
             // leaving the user with a half written file should e.g. the system crash or throw an
-            // error while writing. On Linux we do this by opening the directory for the path
-            // with `O_TMPFILE` and creating a hard link when closing the file. On other platforms
-            // we generate a dot file with a randomised suffix name and rename it to the
+            // error while writing. On non-Android Linux we do this by opening the directory for
+            // the path with `O_TMPFILE` and creating a hard link when closing the file. On other
+            // platforms we generate a dot file with a randomised suffix name and rename it to the
             // destination.
+            #if os(Android)
+            return Self.syncOpenWithMaterialization(
+                atPath: path,
+                mode: mode,
+                options: options,
+                permissions: permissions,
+                executor: executor,
+                useTemporaryFileIfPossible: false
```

Comment:
[No hard links on most Android partitions](https://stackoverflow.com/questions/32365690/is-android-m-not-allowing-hard-links).